complete.py

```diff
@@ -0,0 +1,771 @@
+        if self._write_trajectory:
+            from coddiwomple.particles import Particle
+            self.particle = Particle(0)
+        else:
+            self.particle = None
+        
+    def _before_integration(self, *args, **kwargs):
+        particle_state = args[0] #define the particle state
+        n_iterations = args[1] #define the number of iterations
+        
+        self._current_state_works = [] #define an interim (auxiliary) list that will track the thermodynamic work of the current application
+        self._current_state_works.append(0.0) #the first incremental work is always 0 since the importance function is identical to the first target distribution (i.e. fully interacting MM)
+        
+        self._iteration = 0.0 #define the first iteration as 0
+        self._n_iterations = n_iterations #the number of iterations in the protocol is equal to the number of steps in the application
+        
+        #update the particle state and the particle state subset
+        particle_state.update_from_context(self.context, ignore_velocities=True) #update the particle state from the context
+        self.particle_state_subset = SamplerState(positions = particle_state.positions[list(self._subset_indices_map.keys())]) #create a particle state from the subset context
+        self.particle_state_subset.apply_to_context(self.context_subset, ignore_velocities=True) #apply the subset particle state to its context
+        self.particle_state_subset.update_from_context(self.context_subset, ignore_velocities=True) #update the subset particle state from its context to updated the potential energy
+        #print(f"particle subset reduced potential: {self.pdf_state_subset.reduced_potential(self.particle_state_subset)}")
+           
+        #get the reduced potential
+        reduced_potential = self._compute_hybrid_potential(_lambda = self._iteration / self._n_iterations, particle_state = particle_state)
+        perturbed_reduced_potential = self._compute_hybrid_potential(_lambda = (self._iteration + 1.0) / self._n_iterations, particle_state = particle_state)
+        self._current_state_works.append(self._current_state_works[-1] + (perturbed_reduced_potential - reduced_potential))
+        #print(f"reduced_potential: {reduced_potential}; perturbed_potential: {perturbed_reduced_potential}")
+        
+        #make a new force object
+        mm_force_matrix = self._compute_hybrid_forces(_lambda = (self._iteration + 1.0) / self._n_iterations, particle_state = particle_state).value_in_unit_system(unit.md_unit_system)
+        self.integrator.setPerDofVariableByName('moddi', mm_force_matrix) 
+        
+        #report
+        if self._write_trajectory:
+            self.particle.update_state(particle_state)
+            self.reporter.record([self.particle])
+        
+        #logger
+        #self._log_context_parameters()
+        
+        
+    
+    def _during_integration(self, *args, **kwargs):
+        particle_state = args[0]
```

Comment:
As before, I'm not sure why this is written like this?

Reply:
see above